### PR TITLE
ArchiveHTTP Plugin

### PR DESCRIPTION
**Pull request by GoldBlacky**
I added a plugin called ArchiveHTTP that behaves like ArchiveSF and is meant to replace it. This is because archiveSF is using an API that will become obsolete, and the new API will now be supported through ArchiveHTTP. ArchiveHTTP is not just made to work with the new API but has many changes and improvements to the old ArchiveSF, including general changes to the way multithreaded requests are handled. ArchiverCommon was updated to fix some old bugs, but eventually the changes were too big, so a successor to ArchiverCommon, ArchiverGeneral, was created. This is used in ArchiveHTTP and works from Qt 5.15 onwards. I also cleaned up caCartesianPlot and added some functionality to synchronize zoom. Additionally, some minor bug fixes were made to caQtDM and to ArchiveSF. 
List of Features / Fixes in general caQTDM:
- Updated README with new environment variables
- Fixed memory leaks
- Fixed some general warnings
- Cleaned up redundant classes from caCartesianPlot and caStripPlot into plotHelperClasses.h
- Added functionality to caCartesianPlot to synchronize zoom between plots on both axes and only on the X-axis (time) via signal & slot. (Configurable in designer)
- Improved PlotPicker to work with different time and value scales for the X-axis
- Updated caCalc to also work with QRectF instead of just QRect
- Fixed errors and typos in caCartesianPlot and caStripplot dialogs

List of Fixes in ArchiveSF / archiverCommon:
- Improved some indenting and formatting
- Updated the way monitors are separated and uniquely identified to fix some memory issues
- Fixed edge-case errors when spamming panel reloads
- Cleaned up some code

List of Features / Fixes in ArchiveHTTP compared to ArchiveSF:
- Added ArchiverGeneral, which should replace ArchiverCommon from Qt5.15 onwards.
- Cleaned up the structure and separated classes into their own files
- Reimplemented multithreaded channel requesting mechanism to ensure thread-safety and performance
- Updated the way requests are aborted to ensure no memory leaks occur and no threads keep running unused
- Reimplemented retrieval to reuse data requested by multiple curves and reuse already retrieved data to request sequentially, instead of all at once as before
- Replaced old JSON with QJson due to incredible performance benefits
- Added a class for handling URLs for GET requests to the new API
- Added a class for tracking basic performance data in network requests
- Made general errors more verbose and transparent
- Improved error handling in network operations and made those errors more accurate, verbose, and transparent

### CODE REVIEW

File: caQtDM_Lib/caQtDM_Plugins/archive/archiveSF/archiveSF_plugin.h
Comment (HelgeBrands) [inline caQtDM_Lib/caQtDM_Plugins/archive/archiveSF/archiveSF_plugin.h:1]: same like in the cpp file

File: caQtDM_Lib/caQtDM_Plugins/archive/archiverCommon.cpp
Comment (HelgeBrands) [inline caQtDM_Lib/caQtDM_Plugins/archive/archiverCommon.cpp:1]: same with the changes here, all new things should be made in archiverGeneral

File: caQtDM_Lib/caQtDM_Plugins/archive/archiverCommon.h
Comment (HelgeBrands) [inline caQtDM_Lib/caQtDM_Plugins/archive/archiverCommon.h:1]: same with the changes here, all new things should be made in archiverGeneral

File: caQtDM_Lib/caQtDM_Plugins/controlsinterface.h
Comment (HelgeBrands) [inline caQtDM_Lib/caQtDM_Plugins/controlsinterface.h:1]: Why did you rearrange the includes? Here is only a little bit of cosmetics made no need to change the code

File: caQtDM_Lib/caQtDM_Plugins/modbus/modbus_Plugin.cpp
Comment (HelgeBrands) [inline caQtDM_Lib/caQtDM_Plugins/modbus/modbus_Plugin.cpp:1]: please leave code untouched that you definitely never maintained nor tested

File: caQtDM_QtControls/src/cacartesianplot.cpp
Comment (HelgeBrands) [inline caQtDM_QtControls/src/cacartesianplot.cpp:1]: #if QWT_VERSION >= 0x060100 
There are parts deleted an other parts got an update? What is really needed and what not. Make a real cleanup or not, not a mixture!

File: caQtDM_QtControls/src/plotHelperClasses.h
Comment (HelgeBrands) [inline caQtDM_QtControls/src/plotHelperClasses.h:1]: Please check here the QWT Versions

File: caQtDM_Viewer/src/caQtDM.cpp
Comment (HelgeBrands) [inline caQtDM_Viewer/src/caQtDM.cpp:1]: When you like to change this: Please make this clean, not in this dirty style

File: caQtDM_Viewer/src/fileopenwindow.cpp
Comment (HelgeBrands) [inline caQtDM_Viewer/src/fileopenwindow.cpp:1]: Why did you change this?

File: caQtDM_QtControls/src/cainclude.cpp
Comment (HelgeBrands) [inline caQtDM_QtControls/src/cainclude.cpp:1]: Why did you change this?

**Review by HelgeBrands — COMMENTED** [review]
Please review all comments to check what is needed or explain me the reason for the changes

File: caQtDM_Viewer/src/fileopenwindow.cpp
Comment (GoldBlacky) [reply caQtDM_Viewer/src/fileopenwindow.cpp:1]: Found this as a memory leak while using Valgrind. This also just doesn't serve a purpose, when initializing a MessageWindow(widget), then "widget" is set as the parent widget. This is only done so if the parent is destroyed, the child (here MessageWindow) is also destroyed. However here the "widget" here is never used again so this functionality is not used at all. Even if it were, this would not have any effect, as the messageWindow is added to a gridLayout using "addWidget()" afterwards, which passes ownership to the gridLayout either way, so from then on the parent doesn't do anything anymore expect take up memory. See https://stackoverflow.com/a/8798374 for more information on parent widgets. I also tested this on RH7, RH8, and windows, the fileopenwindow doesn't seem to have any different behaviour, as expected.

File: caQtDM_QtControls/src/cainclude.cpp
Comment (GoldBlacky) [reply caQtDM_QtControls/src/cainclude.cpp:1]: We have discussed this once a while back, this was just a typo where it is written like an alternation, however before the condition didn't matter at all as the curly brackets were missing so the resize was executed no matter the width.
I originally found this because it caused a warning in QtCreator. I tested it before and after using empty includes, nonexistant ones and very small ones, found no difference in behaviour, just one warning less. However, i couldn't manage to actually get 0 as a value because you cannot make an included widget sized 0,0 in the QtDesigner.  Besides that, the function here is never actually called in the whole codebase (no matter the architecture and seems to have been ignored in a bugfix. (I therefore also tested it by calling the function directly). When calling the resize function with 0,0 manually, it simply disappears, but still takes up the space. I am not too sure about this, in the commit where you commented out this function, you wrote it was an incomplete bugfix, maybe we should take a look at it again, the meantioned code is in caqtdm_lib.cpp:8950

File: caQtDM_QtControls/src/cacartesianplot.cpp
Comment (GoldBlacky) [reply caQtDM_QtControls/src/cacartesianplot.cpp:1]: This was just removed because it is a helper class which i copied to plothelperclasses.h, where the "#if QWT_VERSION >= 0x060100" is now, so this was just segregation. Because the castripplot had similar classes that have minor differences but are name equally, I had to change the name of some since they are now in the same header file. Is that what you meant?

File: caQtDM_QtControls/src/cacamera.cpp
[the remark this answers is not in the document]
Comment (GoldBlacky) [reply caQtDM_QtControls/src/cacamera.cpp:1]: I didn't find a messageWindow in caCamera, should I just qDebug the error like this:
if (error != Z_OK) {
    qDebug() << "caCamera: error uncompressing image data";
}
?

File: caQtDM_Lib/caQtDM_Plugins/controlsinterface.h
Comment (GoldBlacky) [reply caQtDM_Lib/caQtDM_Plugins/controlsinterface.h:1]: Yeah, this was because i auto-refactored some files i was currently working on, that automatically rearranged the include, no idea why, will revert it.

File: caQtDM_Lib/caQtDM_Plugins/archive/archive.pro
[the remark this answers is not in the document]
Comment (GoldBlacky) [reply caQtDM_Lib/caQtDM_Plugins/archive/archive.pro:1]: Just discussed this. This works, and I've already tested it. The reason this was done likewise and not as in other already existing parts is because this way it works in all current and future Qt Versions, without the need to update the code. Other versions use e.g. "contains(QT_VER_MAJ, 5)", which would need to be updated with each new version of Qt, if they should also build the conditional block. Instead of a numerical comparison, as it is done in many other parts, this is a string comparison, where QT_VERSION is in the format I used here. (Numerical comparisons are not really supported on Qt4, which is why other pieces use this in conjunction with e.g. "contains(QT_VER_MAJ, 5)")
I have tested it on RH7, RH8, Windows and also by manually setting alues for QT_VERSION. 

Normally, QT_VERSION does not have this format, but because it is split in qtdefs.pri it was modified to have this format:
image

File: caQtDM_Lib/caQtDM_Plugins/archive/archiveSF/archiveSF_plugin.cpp
[the remark this answers is not in the document]
Comment (GoldBlacky) [reply caQtDM_Lib/caQtDM_Plugins/archive/archiveSF/archiveSF_plugin.cpp:1]: When I started, I noticed some errors in archiverCommon which would lead to crashes in Edge-Cases (always have, just haven't been discovered yet), so i fixed them and also changed some other things while on it. In this file, the changes are only one memory leak via valgrind (delete archiverCommon), and some pretty reasonable renames, e.g. from myThread to workerSfThread.

File: caQtDM_Lib/caQtDM_Plugins/archive/archiveSF/archiveSF_plugin.h
Comment (GoldBlacky) [reply caQtDM_Lib/caQtDM_Plugins/archive/archiveSF/archiveSF_plugin.h:1]: myThread -> WorkerSfThread is pretty reasonable I think, the reason I started changing this file is described in the comment for the header file.

File: caQtDM_Lib/caQtDM_Plugins/archive/archiverCommon.cpp
Comment (GoldBlacky) [reply caQtDM_Lib/caQtDM_Plugins/archive/archiverCommon.cpp:1]: Also see the comment I wrote on the archiveSF_plugin.cpp. Here I noticed some errors which can and did lead to crashes in edge cases such as reloading widgets quickly or with certain timings. Also, I believe there werde some multithreaded & memory issues. So I only did some bugfixing here, which I also tested with archiveSF. Besides that, I also did some refactoring as I was on it. I left those bugfixes because it is still fully compatible with ArchiveSF (generally all other archivers, as the class interface doesnt change in a way which removes or changes functionality for archivers which use the archiverCommon).I can revert those changes if you want, this will just bring back some issues we previously had with ArchiveSF.

**Comment by GoldBlacky** [comment]
I reverted changes that are not relevant for this PR, will open another PR for them afterwards.
